pkg/kubeconfig/kubeconfig.go

```diff
@@ -25,15 +29,46 @@ func FromEnvironment() string {
 	return os.Getenv(EnvName)
 }
 
-type missingFileError struct {
-	Path string
+type Validator interface {
+	ValidateFile(filename string) error
+	Validate(r io.Reader) error
+}
+
+type validator struct {
+	load func([]byte) (*clientcmdapi.Config, error)
+}
```

Comment:
Have we considered modelling this as just a function (`func Validate(...)`)? This feels quite heavy and I'm not sure we really need an object at consumer sites.

Reply:
Oh boy, have we! Let me tell you, the permutations we've been through.

I tend to agree, that this is all quite over-engineered at this point. I'd say I don't know how it got this far, but that's not true, cuz I was there for every step.

So, to your question, I did have a Validate function for a while. However, in order to test it, I had to be able to mock out the call to clientcmd.Load(), and to do that with a function, I had to use a package-level variable. This wasn't ideal, as it wasn't thread-safe. Nevermind that every one of the EKS-A CLI commands does the same (is that normal with cobra.Command's, and if so, why does anyone use that package?)

So to try to remove the package-level variable, I added a parameter to Validate that let one specify a validator to use. That led to a discussion around the use of nil vs variadic parameters, which lead to a bunch of interfaces and then eventually ended up ... here.

I guess a lot of that discussion ended up happening over Slack for speed's sake. It might have been interesting if it had been captured in this PR. Well, interesting in some weird academic meets train wreck sort way.

Reply:
Isn't loading the configuration the crux of validation? Without it, any testing seems rather redundant.

I think we should go back to the beginning of the road and just have a `func Validate(io.Reader) error`. Creating invalid configurations should be trivial as would generating an io error and we'd actually test that a `kubeconfig.Validate(reader)` functions as expected. Given this is generally serving main type functions that are operating on path strings I would even suggest `Validate(path string) error` and only introduce the `io.Reader` when its needed. Using the path simplifies our function further to just `clientcmd.LoadFromFile(...)`.

I do empathize with a desire to push the `io` package but given the only consumer is a main type func right now it doesn't seem appropriate in this case.

Sorry to step in last minute here, but I don't think what we have in the current iteration is the right approach.

Reply:
Yeah, that seems like a good way to remove a fair bit of cruft.